armi/reactor/reactors.py

```diff
@@ -2262,42 +2251,54 @@ def processLoading(self, cs):
+                    self.parent.blueprints.assemblies.values(),
+                    key=lambda a: len(a),
+                    reverse=True,
+                )[0]
+                for a in self.parent.blueprints.assemblies.values():
+                    a.makeAxialSnapList(refAssem=finestAssemblyMesh)
+
+        else:
             self.p.referenceBlockAxialMesh = self.findAllAxialMeshPoints(
                 applySubMesh=False
             )
-        if not self.p.axialMesh:
             self.p.axialMesh = self.findAllAxialMeshPoints()
-
-        refAssem = self.refAssem
-        # blueprints.assemblies.values need to be populated
-        # In a load from DB case construction may not have been prepped yet.
-        # this normally happens during blueprints constructAssem, but for DB
-        # load this is not called.
-        self.parent.blueprints._prepConstruction(cs)
-        if not cs["detailedAxialExpansion"]:
-            # prepare core for mesh snapping during axial expansion
-            for a in self.getAssemblies():
-                # No BOL assemblies since the core may have an expanded mesh
-                # That wont work for snap list. BOL (blueprints) assemblies
-                # will have the expansion applied upon core.createAssemblyOfType
-                a.makeAxialSnapList(refAssem=refAssem)
-
-            # Now apply mesh snapping for blueprints
-            # All mesh points should line up with the ref mesh, so first
-            # find the assembly with the finest mesh (it is the ref).
-            finestMeshA = sorted(
-                self.parent.blueprints.assemblies.values(),
-                key=lambda a: len(a),
-                reverse=True,
-            )[0]
-            # Apply an unexpanded ref mesh to the blueprints.Since they are
-            # not yet expanded, the normal core mesh would not line up.
-            # They will get expanded by core.createAssemblyOfType.
-            for a in self.parent.blueprints.assemblies.values():
-                a.makeAxialSnapList(refAssem=finestMeshA)
+            if not cs["detailedAxialExpansion"]:
+                # prepare core for mesh snapping during axial expansion
+                for a in self.getAssemblies(includeAll=True):
+                    a.makeAxialSnapList(self.refAssem)
+
+            if not cs["inputHeightsConsideredHot"]:
+                runLog.header(
+                    "=========== Axially expanding all (except control) assemblies from Tinput to Thot ==========="
+                )
+                axialExpChngr = AxialExpansionChanger(cs["detailedAxialExpansion"])
+                for a in self.getAssemblies():
+                    if not a.hasFlags(Flags.CONTROL):
+                        axialExpChngr.setAssembly(a)
+                        axialExpChngr.expansionData.computeThermalExpansionFactors()
+                        axialExpChngr.axiallyExpandAssembly(thermal=True)
+                axialExpChngr._manageCoreMesh(self.parent)
```

Comment:
this seems like it needs to be a public method if called in this way.

Do you think it would be reasonable to move 2295-2301 into a new method AxialExpansionChanger  called `applyThermalExpansion` or something like that so that there is an easy way to apply it and you don't have to worry about calling the manageMesh method?

Reply:
> this seems like it needs to be a public method if called in this way.

Good catch. Changing `_manageCoreMesh` from private to public was coming in #787 , but will be included here as well. In https://github.com/terrapower/armi/pull/797/commits/7df5b13b56fe5ec2afb5ffa07487001a816833d2 I change it to be a public facing method.

> Do you think it would be reasonable to move 2295-2301 into a new method AxialExpansionChanger called applyThermalExpansion or something like that so that there is an easy way to apply it and you don't have to worry about calling the manageMesh method?

So this is certainly possible - and actually sort of exists in `axiallyExpandCoreThermal`, but is getting removed in #787. Having a core level axial expansion functionality is useful, but I found that it doesn't support general applications very well. E.g., here we want to do thermal expansion based on a Delta T between `Tinput` and `Thot`, but elsewhere someone may want to do thermal expansion based on a different Delta T (e.g., a new `Thot` and previous `Thot`). So then the method we would write for one case would require another method to support it - and core level methods are particularly heavy to get unit testing coverage for. We might(?) be able to configure a core level function that better supports generality, but it would still be really heavy to test.